src/GovUk.Education.ExploreEducationStatistics.Admin/Services/Public.Data/PreviewTokenService.cs

```diff
@@ -0,0 +1,101 @@
+using GovUk.Education.ExploreEducationStatistics.Common.Model;
+using GovUk.Education.ExploreEducationStatistics.Common.Services.Interfaces.Security;
+using GovUk.Education.ExploreEducationStatistics.Common.Validators;
+using GovUk.Education.ExploreEducationStatistics.Common.Validators.ErrorDetails;
+using GovUk.Education.ExploreEducationStatistics.Common.ViewModels;
+using GovUk.Education.ExploreEducationStatistics.Content.Model.Database;
+using GovUk.Education.ExploreEducationStatistics.Public.Data.Model;
+using GovUk.Education.ExploreEducationStatistics.Public.Data.Model.Database;
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.EntityFrameworkCore;
+using ValidationMessages = GovUk.Education.ExploreEducationStatistics.Admin.Validators.ValidationMessages;
+
+namespace GovUk.Education.ExploreEducationStatistics.Admin.Services.Public.Data;
+
+public class PreviewTokenService(
+    ContentDbContext contentDbContext,
+    PublicDataDbContext publicDataDbContext,
+    IUserService userService
+) : IPreviewTokenService
+{
+    public async Task<Either<ActionResult, PreviewTokenViewModel>> CreatePreviewToken(
+        Guid dataSetVersionId, string label, CancellationToken cancellationToken = default)
+    {
+        return await userService.CheckIsBauUser()
+            .OnSuccess(async () => await CheckDataSetVersionExists(dataSetVersionId, cancellationToken))
+            .OnSuccessDo(HandleDataSetVersionNotDraft)
+            .OnSuccess(async () =>
+            {
+                var previewToken = publicDataDbContext.PreviewTokens.Add(new PreviewToken
+                {
+                    DataSetVersionId = dataSetVersionId,
+                    Label = label,
+                    Expiry = DateTimeOffset.UtcNow.AddDays(1),
+                    CreatedByUserId = userService.GetUserId()
+                });
+                await publicDataDbContext.SaveChangesAsync(cancellationToken);
+                return await MapPreviewToken(previewToken.Entity);
+            });
+    }
+
+    public async Task<Either<ActionResult, PreviewTokenViewModel>> GetPreviewToken(
+        Guid previewTokenId, CancellationToken cancellationToken = default)
+    {
+        return await userService.CheckIsBauUser()
+            .OnSuccess(async () => await publicDataDbContext.PreviewTokens
+                .AsNoTracking()
+                .SingleOrNotFoundAsync(pt => pt.Id == previewTokenId, cancellationToken: cancellationToken))
+            .OnSuccess(MapPreviewToken);
+    }
+
+    private async Task<Either<ActionResult, DataSetVersion>> CheckDataSetVersionExists(
+        Guid dataSetVersionId,
+        CancellationToken cancellationToken)
+    {
+        return await publicDataDbContext.DataSetVersions
+            .AsNoTracking()
+            .SingleOrNotFoundAsync(dsv => dsv.Id == dataSetVersionId, cancellationToken);
+    }
+
+    private static Either<ActionResult, Unit> HandleDataSetVersionNotDraft(DataSetVersion dataSetVersion)
```

Comment:
Would `ValidateDraftDataSetVersion` be more in keeping with other service validation methods?  I honestly don't know!

Reply:
It sounds better to me, I've made this change, thanks.